### ISSUE TITLE
Possible driver timeouts or reduced performance after #10540 (seen on Skate 3)

### ISSUE THREAD

**Issue by schm1dtHL** [RPCS3/rpcs3#10568]
[Log1.log](https://github.com/RPCS3/rpcs3/files/6835276/Log1.log)
[Log2.log.gz](https://github.com/RPCS3/rpcs3/files/6835312/Log2.log.gz)
[Log3.log.gz](https://github.com/RPCS3/rpcs3/files/6835311/Log3.log.gz)

## Quick summary
When running RPCS3 on an iGPU, in my case the one included with the Ryzen 5 5600H, a driver timeout can occur when booting Skate 3 from my testing, although this may affect other games too. If a timeout doesn't occur, it would CTD prior to #10565, after which it would boot, but would exhibit worse performance compared to builds prior to #10540.

## Details
Settings used were defaults, except for the following:
Logs 1,2 and 3: Anti-Aliasing - Disabled(Improves performance, recommended by the wiki)
Logs 2 and 3: Thread Scheduler - Alternative(Improves performance, OS Scheduler was tested too and exhibited the same issues)
Logs 2 and 3: Max SPURS Threads - 4(Improves performance, Unlimited was tested too and exhibited the same issues)

Log1 is from a run which led to a driver timeout(post #10565),
Log2 is from a run which went ingame and exhibited performance issues(post #10565),
Log3 is from a run which did not exhibit either issues(pre #10540).

Pre-regression performance was ~46-48fps, post-regression perf is ~33fps. ( Both were in fullscreen mode )
iGPU memory usage in the cases of Log2 and Log3 were both at ~284MB dedicated and 1GB shared, although this can vary.

Screenshots: Unable to obtain them for Log2 and Log3 as of now due to fullscreen and screenshot tools not cooperating, whenever I can do so they'll be added to the ticket. Log1 has no useful info from a screenshot due to the driver timeout.

* OS - Windows 10 Home 21H1
* CPU - Ryzen 5 5600H
* GPU - AMD Radeon Graphics( Vega ), an RTX 3050 Ti is present but is not used in the testing for this issue.
* Driver version - 21.6.1 for AMD Radeon, 471.11 for Nvidia.

**Comment by kd-11** [comment]
This is a quirk of AMD hardware. There are always 2 VRAM heaps. One is 256M and special (host_coherent + device_local) and is present in all their hw. This means 4GB VRAM ends up being around 3.75GB primary and 256M secondary.
Things get nasty when working with a total pool of 512M, because now you have 256M+256M. We only use the primary heap so we're working with a 256M budget only.
Now, each application is assigned a memory budget by the driver (you don't actually have access to 100% VRAM) and from the slow heap I can see we're waaaay over budget there, meaning we're competing with other applications.
I'll see if I can do some kind of heap abstraction to merge heaps0 and 2 and gain that extra 256M of VRAM, though realistically we're right at the edge of what can even be supported by rpcs3.

**Comment by kd-11** [comment]
@schm1dtHL How much dedicated VRAM total does task manager show? Does it show 256M or 512M? I ask because there is a chance the 2 heaps alias and allocating from one just subtracts from the other which would have very different implications.

**Comment by schm1dtHL** [comment]
Task manager shows a total of 512M dedicated for the iGPU.

**Comment by kd-11** [comment]
@schm1dtHL Please try with https://github.com/RPCS3/rpcs3/pull/10604 once an artifact is ready. If you still encounter problems attach the log file. I have tried to combine all vram pools if detected so that we should use the 512MB total instead of the separate 256MB pieces. There is still some code regarding detecting when we need to actually take action, but this requires that I roll out some custom code for some things, so it will come later.

**Comment by schm1dtHL** [comment]
Tested the PR, performance is closer to pre-regression perf now, being at around 40-41fps. Driver timeouts appear to still occur sometimes though.

**Comment by kd-11** [comment]
Hmm, so we still have lost 5-7 fps. Please attach the new log.

**Comment by schm1dtHL** [comment]
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/6871480/RPCS3.log.gz)

**Comment by kd-11** [comment]
There is no sign of slowdown in the new log.

**Comment by kd-11** [comment]
Task List:
- [x] Pool together similar memory types instead of simply picking the 'largest' heap.
- [x] Rebalance load between the pooled heaps to avoid loading one heap more than the others.
- [ ] Fix use-after-free for surface cache references where the referenced item may have been destroyed due to memory pressure.

**Comment by kd-11** [comment]
https://github.com/RPCS3/rpcs3/pull/10611 should solve the remaining performance issues. 
Attach log if it does not (get to the parts with degraded performance and let it sit there for a 30 seconds, then provide the log file)

I know why the crashes happen with driver timeout, I'm working on the fix.

**Comment by schm1dtHL** [comment]
Performance is the same as before, let it sit as you said to ensure that it hadn't changed in a significant way. Log is attached below, it includes a run with the 3050 Ti first,  then the iGPU run second, the former was just done to create a new shader cache and SPU cache since not having either when running with the iGPU guarantees a timeout(I cleared those caches to try ensuring that they wouldn't interfere with the tests) 

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/6881209/RPCS3.log.gz)

**Comment by kd-11** [comment]
@schm1dtHL I should have fixed the crashing now with the latest commits in that PR, you can retest that at least.
As for performance, the log shows that there is no 'panic' in the driver now since it can see the full 512M now. This is good, but I cannot see why performance would be still be lower than before. In fact, the log from the older build shows much more activity that would cause slowdown, so I'm surprised it is still that much faster. Is the 3050Ti also affected by this performance regression? Maybe its worth retesting in case windows is creating memes. Run the test with both exes having the same file name just in case to eliminate driver issues. I'll try and get more people to test the before and after as well.

**Comment by schm1dtHL** [comment]
Will do so tomorrow, regarding the 3050 Ti though, whilst I haven't noticed a regression with it, I have consistently noticed performance on it that's weaker than what the iGPU gets, both prior to the memory management changes and after them(it gets about 30fps in the same tested area as what the iGPU has been tested on).

**Comment by kd-11** [comment]
Closing for now, If we are still able to observe degraded performance after the memory management was added, I will reopen and we can do a side investigation to figure out why performance got worse.

**Comment by schm1dtHL** [comment]
Retested, indeed the regressions appear to be fixed now on the latest master builds, with the game running at ~50-51 fps in that area now with the iGPU.

**Comment by kd-11** [comment]
Thanks for confirming. The 3050Ti's poor performance is indeed strange, but can sometimes be simply explained by the fact that the iGP uses system RAM where the emulated VM also lives cutting out the 'middle-man' (PCI-e). This is typically only seen on old motherboards where the GPU may not have full access to at least PCI-e 3.0 x16 with fully active lanes which is common when a NVME SSD is present or other peripherals that use up PCI-e lanes.

**Comment by schm1dtHL** [comment]
Interesting, would it be worth making a new issue for the 3050 Ti perf issues? It does run with PCIe 3.0 x8, but I think that shouldn't lead to it performing way worse than the iGPU in RPCS3's case(30-33fps compared to the IGP's ~50-51fps in that Skate 3 area).

**Comment by kd-11** [comment]
Unless it is something we can fix in code, there is no reason to open a ticket for it. Rpcs3 is very bandwidth intensive overall, there is not much we can do about that due to how the PS3 works, at least not at the moment.
If you are using MSAA, try turning it off to save performance. This is a known issue where MSAA can have a much greater performance impact on NVIDIA cards.

**Comment by schm1dtHL** [comment]
RIP, the prior tests were with MSAA off on both the dGPU and iGPU, would guess that MSAA on would lead to worse disparities. Thanks for helping out with the iGPU regressions though.
